include/vinecopulib/bicop/implementation/kernel.ipp

```diff
@@ -16,50 +17,80 @@ inline KernelBicop::KernelBicop()
   Eigen::VectorXd grid_points(m);
   for (size_t i = 0; i < m; ++i)
     grid_points(i) = -3.25 + i * (6.5 / static_cast<double>(m - 1));
-  grid_points = tools_stats::pnorm(grid_points);
-  
-  // move boundary points to 0/1, so we don't have to extrapolate
-  grid_points(0) = 0.0;
-  grid_points(m - 1) = 1.0;
-  
   interp_grid_ = std::make_shared<tools_interpolation::InterpolationGrid>(
-    grid_points,
+    tools_stats::pnorm(grid_points),
     Eigen::MatrixXd::Constant(m, m, 1.0) // independence
   );
 }
 
 inline Eigen::VectorXd
-KernelBicop::pdf_raw(const Eigen::Matrix<double, Eigen::Dynamic, 2>& u)
+KernelBicop::pdf_raw(const Eigen::MatrixXd& u)
+{
+  auto pdf = interp_grid_->interpolate(u);
+  auto trunc = [] (const double& p) {
+      return std::fmax(p, 1e-20);
+  };
+  return tools_eigen::unaryExpr_or_nan(pdf, trunc);
+}
+
+inline Eigen::VectorXd
+KernelBicop::pdf(const Eigen::MatrixXd& u)
 {
-  return interp_grid_->interpolate(u);
+  if (u.cols() == 4) {
+    // evaluate jittered density at mid rank for stability
```

Comment:
Can you explain? Why do we treat the kernel differently from the others?

Reply:
(same question for the hfuncs/hinvs)

Reply:
Let's talk about the purely discrete case for simplicity. The copula "pdf" is 
> p(u, v) := C(u, v) - C(u^-, v) - C(u, v^-) + C(u^-, v^-)

Now, the (true) jittered density is constant in the square [u^-,u] x [v^-, v] with value p(u, v). If we estimate it, we will have the least bias in the center of this square. That's what I compute in the specialized pdf. Computing the sum/difference of the cdfs would be computationally more demanding and less accurate.

The reasoning is the same for hfuncs and discrete-continuous pairs.